x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/TransportPutShutdownNodeAction.java

```diff
@@ -49,8 +56,42 @@ protected void masterOperation(
         ClusterState state,
         ActionListener<AcknowledgedResponse> listener
     ) throws Exception {
-        // TODO: implement me!
-        listener.onResponse(AcknowledgedResponse.of(true));
+        // Verify that the requested node actually exists in the cluster
+        if (state.nodes().nodeExists(request.getNodeId()) == false) {
+            throw new IllegalArgumentException("there is no node with id [" + request.getNodeId() + "] in this cluster");
+        }
+
+        clusterService.submitStateUpdateTask(
+            "put-node-shutdown-" + request.getNodeId(),
+            new AckedClusterStateUpdateTask(request, listener) {
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    NodesShutdownMetadata currentShutdownMetadata = currentState.metadata().custom(NodesShutdownMetadata.TYPE);
+                    if (currentShutdownMetadata == null) {
+                        currentShutdownMetadata = new NodesShutdownMetadata(new HashMap<>());
+                    }
+
+                    // Verify that there's not already a shutdown metadata for this node
+                    if (Objects.nonNull(currentShutdownMetadata.getNodeMetadata(request.getNodeId()))) {
+                        throw new IllegalArgumentException("node [" + request.getNodeId() + "] is already shutting down");
+                    }
```

Comment:
Should we allow a change in type or reason for a shutdown? I think it'd be nice to be able to change a `restart` type to `removal` if we decide that we wanted to remove the node after all.

For now though, we can keep this and maybe think about revisiting it in the future, what do you think?

Reply:
Yes, we've discussed being able to change the type, at least in one direction. That's a pretty easy follow-up though, so I think we should move forward with this and add that functionality when we need it.

I opened https://github.com/elastic/elasticsearch/issues/71305 to keep track.